packages/codec/lib/abi-data/allocate/index.ts

```diff
@@ -972,6 +1061,122 @@ export function getCalldataAllocations(
   return allocations;
 }
 
+function getReturndataAllocationsForContract(
+  abi: Abi.Abi,
+  contractNode: Ast.AstNode | undefined,
+  referenceDeclarations: Ast.AstNodes,
+  userDefinedTypes: Format.Types.TypesById,
+  abiAllocations: AbiAllocations,
+  compilationId: string,
+  compiler: Compiler.CompilerVersion | undefined
+): RevertReturndataAllocation[] {
+  let useAST = Boolean(contractNode && contractNode.usedErrors);
+  if (useAST) {
+    const errorNodes = contractNode.usedErrors.map(
+      errorNodeId => referenceDeclarations[errorNodeId]
+    );
+    let abis: Abi.ErrorEntry[];
+    try {
+      abis = errorNodes.map(
+        errorNode => <Abi.ErrorEntry>Ast.Utils.definitionToAbi(
+          errorNode,
+          referenceDeclarations
+        )
+      );
+    } catch {
+      useAST = false;
+    }
+    if (useAST) { //i.e. if the above operation succeeded
+      return contractNode.usedErrors.map(
+        errorNodeId => referenceDeclarations[errorNodeId]
+      ).map((errorNode, index) => allocateError(
+        abis[index],
+        errorNode,
+        referenceDeclarations,
+        userDefinedTypes,
+        abiAllocations,
+        compilationId,
+        compiler
+      ));
+    }
+  }
+  if (!useAST) { //deliberately *not* an else!
+    return abi
+      .filter((abiEntry: Abi.Entry) => abiEntry.type === "error")
+      .filter(
+        (abiEntry: Abi.ErrorEntry) =>
+          !AbiDataUtils.abiEntryIsObviouslyIllTyped(abiEntry)
+      ) //hack workaround
+      .map((abiEntry: Abi.ErrorEntry) => allocateError(
+        abiEntry,
+        undefined,
+        referenceDeclarations,
+        userDefinedTypes,
+        abiAllocations,
+        compilationId,
+        compiler
+      ));
+  }
+}
```

Comment:
Can this be cleaned up to avoid mutating `useAST`? It's kind of difficult to look at

Reply:
Hm, I mean, it could if we introduce additional function calls?  Like a function for the AST case and a function for the ABI case?  I don't really want to break things up even further like that, but I could if you think that's warranted.  Note we already do this sort of thing elsewhere in allocation for handling falling back from full mode to ABI mode.

Reply:
Like to be clear I don't want to break this up with further function calls because we have this repeating parallel 3-level structure, this function isn't very large so it's more confusing to break it up and have to scroll around rather than being able to just see it all together.

Reply:
Eh we're just going to need to clean things up in the future I guess.